File: c_src/epx_gc.c
/***************************************************************************
 *
 * Copyright (C) 2007 - 2012, Rogvall Invest AB, <tony@rogvall.se>
 *
 * This software is licensed as described in the file COPYRIGHT, which
 * you should have received as part of this distribution. The terms
 * are also available at http://www.rogvall.se/docs/copyright.txt.
 *
 * You may opt to use, copy, modify, merge, publish, distribute and/or sell
 * copies of the Software, and permit persons to whom the Software is
 * furnished to do so, under the terms of the COPYRIGHT file.
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
 * KIND, either express or implied.
 *
 ***************************************************************************/
//
// EPX GC functions
//

#include <stdlib.h>
#include "epx_colors.h"
#include "epx_object.h"
#include "epx_gc.h"

epx_gc_t epx_default_gc =
{
    {0, 0},
    0,            // Not on heap
    0,            // No references
    0,            // No release function
    0,            // Opaque
    EPX_GC_TYPE,  // Type
    0,            // Next

    EPX_FILL_STYLE_SOLID,  // fill_style
    EPX_PIXEL_RED,             // fill_color
    0,                      // fill_texture

    EPX_LINE_STYLE_SOLID,  // line_style
    EPX_JOIN_STYLE_MITER,  // line_join_style
    EPX_CAP_STYLE_NONE,    // line_cap_style
    1,                     // line_width
    0,                     // line_texture

    EPX_BORDER_STYLE_SOLID, // border_style
    EPX_JOIN_STYLE_MITER,   // border_join_style
    EPX_CAP_STYLE_NONE,     // border_cap_style
    EPX_PIXEL_BLACK,        // border_color
    0,                      // border_width
    0,                      // border_texture

    // Color
    EPX_PIXEL_BLUE,             // forground_color
    EPX_PIXEL_GREEN,            // background_color
    // Alpha
    255,                     // alpha factor = 1.0
    // Text
    0,                       // Font used for text drawing
    // Glyphs
    0,                       // glyph_delta_x
    0,                       // glyph_delta_y
    0,                       // glyph_fixed_width
    0                        // glyph_dot_kern
};


void epx_gc_init(epx_gc_t* gc)
{
    EPX_OBJECT_INIT(gc, EPX_GC_TYPE);

    gc->fill_style        = EPX_FILL_STYLE_SOLID;
    gc->fill_color        = epx_pixel_red;
    gc->fill_texture      = 0;

    gc->line_style        = EPX_LINE_STYLE_SOLID;
    gc->line_join_style   = EPX_JOIN_STYLE_MITER;
    gc->line_cap_style    = EPX_CAP_STYLE_NONE;
    gc->line_width        = 1;
    gc->line_texture      = 0;

    gc->border_style      = EPX_BORDER_STYLE_SOLID;
    gc->border_join_style = EPX_JOIN_STYLE_MITER;
    gc->border_cap_style  = EPX_CAP_STYLE_NONE;
    gc->border_color      = epx_pixel_black;
    gc->border_width      = 0;
    gc->border_texture    = 0;

    gc->foreground_color = epx_pixel_blue;
    gc->background_color = epx_pixel_green;

    gc->fader_value      = ALPHA_FACTOR_1;  // 1.0

    gc->font = 0;

    gc->glyph_delta_x     = 0;
    gc->glyph_delta_y     = 0;
    gc->glyph_fixed_width = 0;
    gc->glyph_dot_kern    = 0;
}

epx_gc_t* epx_gc_create()
{
    epx_gc_t* gc;

    if (!(gc = (epx_gc_t*) malloc(sizeof(epx_gc_t))))
	return 0;
    epx_gc_init(gc);
    gc->on_heap = 1;
    gc->refc = 1;
    return gc;
}

void epx_gc_destroy(epx_gc_t* gc)
{
    epx_object_unref(gc);
}

void EPX_GC_TYPE_RELEASE(void* arg)
{
    epx_gc_t* gc = (epx_gc_t*) arg;

    EPX_DBGFMT_MEM("EGC_TYPE_RELEASE: %p", arg);
    epx_object_unref(gc->fill_texture);
    epx_object_unref(gc->line_texture);
    epx_object_unref(gc->border_texture);
    epx_object_unref(gc->font);
    if (gc->on_heap) free(gc);
}

void epx_gc_init_copy(epx_gc_t* src, epx_gc_t* dst)
{
    *dst = *src;
    epx_object_ref(dst->fill_texture);
    epx_object_ref(dst->line_texture);
    epx_object_ref(dst->border_texture);
    epx_object_ref(dst->font);
    dst->next = 0;
}

epx_gc_t* epx_gc_copy(epx_gc_t* gc)
{
    epx_gc_t* copy;

    if (!(copy = (epx_gc_t*) malloc(sizeof(epx_gc_t))))
	return 0;
    // reference textures etc 
    epx_gc_init_copy(gc, copy);
    copy->refc    = 1;
    copy->on_heap = 1;
    return copy;
}

File: src/epx_gc.erl
%%%---- BEGIN COPYRIGHT --------------------------------------------------------
%%%
%%% Copyright (C) 2007 - 2012, Rogvall Invest AB, <tony@rogvall.se>
%%%
%%% This software is licensed as described in the file COPYRIGHT, which
%%% you should have received as part of this distribution. The terms
%%% are also available at http://www.rogvall.se/docs/copyright.txt.
%%%
%%% You may opt to use, copy, modify, merge, publish, distribute and/or sell
%%% copies of the Software, and permit persons to whom the Software is
%%% furnished to do so, under the terms of the COPYRIGHT file.
%%%
%%% This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
%%% KIND, either express or implied.
%%%
%%%---- END COPYRIGHT ----------------------------------------------------------
%%% File    : epx_gc.erl
%%% Author  : Tony Rogvall <tony@PBook.local>
%%% Description : EPX GC process wrapper
%%% Created : 24 Oct 2006 by Tony Rogvall <tony@PBook.local>

-module(epx_gc).

-export([create/0, clear/0, copy/1]).
-export([push/0, pop/0, current/0]).
-export([draw/1]).

-export([set_item/2, set_item/3]).
-export([set/1, set/2]).
-export([get_item/1, get_item/2]).
-export([get/1, get/2]).

-export([set_background_color/2, set_background_color/1,
	 set_foreground_color/2, set_foreground_color/1,
	 set_fill_style/2,	 set_fill_style/1,
	 set_fill_color/2,	 set_fill_color/1,
	 set_fill_texture/2,	 set_fill_texture/1,

	 set_line_style/2,	  set_line_style/1,
	 set_line_width/2,	  set_line_width/1,
	 set_line_join_style/2,	  set_line_join_style/1,
	 set_line_cap_style/2,	  set_line_cap_style/1,
	 set_line_texture/2,	  set_line_texture/1,

	 set_border_style/2,	  set_border_style/1,
	 set_border_color/2,	  set_border_color/1,
	 set_border_width/2,	  set_border_width/1,
	 set_border_join_style/2, set_border_join_style/1,
	 set_border_cap_style/2,  set_border_cap_style/1,
	 set_border_texture/2,	  set_border_texture/1,
	 set_font/2,              set_font/1
	]).

-export([get_background_color/1,  get_background_color/0,
	 get_foreground_color/1,  get_foreground_color/0,
	 get_fill_style/1,	  get_fill_style/0,
	 get_fill_color/1,	  get_fill_color/0,
	 get_fill_texture/1,	  get_fill_texture/0,

	 get_line_style/1,	  get_line_style/0,
	 get_line_width/1,	  get_line_width/0,
	 get_line_join_style/1,	  get_line_join_style/0,
	 get_line_cap_style/1,	  get_line_cap_style/0,
	 get_line_texture/1,	  get_line_texture/0,

	 get_border_style/1,	  get_border_style/0,
	 get_border_color/1,	  get_border_color/0,
	 get_border_width/1,	  get_border_width/0,
	 get_border_join_style/1, get_border_join_style/0,
	 get_border_cap_style/1,  get_border_cap_style/0,
	 get_border_texture/1,	  get_border_texture/0,
	 get_font/1,              get_font/0
	]).

-define(CURRENT, '_EPX_GC_CURRENT').
-define(STACK,   '_EPX_GC_STACK').

create() ->
    epx:gc_create().

copy(Gc) ->
    epx:gc_copy(Gc).

draw(Fun) ->
    %%  Should implement copy on write!
    push(),
    %% Do NOT pop inside fun, unless intensional ;-)
    try Fun() of
	Res -> Res
    catch
	error:Reason ->
	    io:format("draw: crashed, ~p\n", [erlang:get_stacktrace()]),
	    {error,Reason};
	exit:Reason ->
	    io:format("draw: exit, reason = ~p\n", [erlang:get_stacktrace()]),
	    {error,Reason};
	thrown:Value ->
	    Value
    after 
	pop()
    end.

current() ->
    case erlang:get(?CURRENT) of
	undefined ->
	    Gc = create(),
	    put(?CURRENT, Gc),
	    put(?STACK, []),
	    Gc;
	Gc -> Gc
    end.

%% clear EPX_GC stack
clear() ->
    erase(?CURRENT),
    erase(?STACK).


push() ->
    Gc = current(),
    GcCopy = copy(Gc),  %% fixme copy on write
    Stack = erlang:get(?STACK),
    put(?STACK, [Gc|Stack]),
    put(?CURRENT, GcCopy),
    ok.

pop() ->
    [Gc|Stack] = erlang:get(?STACK),
    put(?CURRENT, Gc),
    put(?STACK, Stack),
    ok.

%%
%% General colors
%%
set_background_color(Color) -> set_background_color(current(),Color).
set_background_color(Gc,Color) -> set_item(Gc,background_color, Color).

set_foreground_color(Color) -> set_foreground_color(current(),Color).
set_foreground_color(Gc,Color) -> set_item(Gc,foreground_color, Color).

%%
%% FILL style/color/texture
%%
set_fill_style(Style) ->  set_fill_style(current(),Style).
set_fill_style(Gc,Style) -> set_item(Gc, fill_style, Style).

set_fill_color(Color) -> set_fill_color(current(),Color).
set_fill_color(Gc,Color) -> set_item(Gc, fill_color, Color).

set_fill_texture(Texture) -> set_fill_texture(current(),Texture).
set_fill_texture(Gc,Texture) -> set_item(Gc, fill_texture, Texture).

%%
%% BORDER style/color/join/cap/texture
%%
set_border_style(Style) ->  set_border_style(current(), Style).
set_border_style(Gc,Style) -> set_item(Gc, border_style, Style).

set_border_color(Color) -> set_border_color(current(),Color).
set_border_color(Gc,Color) -> set_item(Gc, border_color, Color).

set_border_width(Width) -> set_border_width(current(), Width).
set_border_width(Gc,Width) when is_integer(Width), Width >= 0  ->
    set_item(Gc, border_width, Width).

set_border_join_style(Style) -> set_border_join_style(current(),Style).
set_border_join_style(Gc,Style) -> set_item(Gc, border_join_style, Style).

set_border_cap_style(Style) -> set_border_cap_style(current(),Style).
set_border_cap_style(Gc,Style) -> set_item(Gc, border_cap_style, Style).

set_border_texture(Texture) -> set_border_texture(current(),Texture).
set_border_texture(Gc,Texture) -> set_item(Gc, border_texture, Texture).

%%
%% LINE style/join/cap/texture
%%
set_line_style(Style) ->  set_line_style(current(), Style).
set_line_style(Gc,Style) -> set_item(Gc, line_style, Style).

set_line_width(Width) -> set_line_width(current(), Width).
set_line_width(Gc,Width) when is_integer(Width), Width >= 0 ->
    set_item(Gc, line_width, Width).

set_line_join_style(Style) -> set_line_join_style(current(),Style).
set_line_join_style(Gc,Style) -> set_item(Gc, line_joine_style, Style).

set_line_cap_style(Style) -> set_line_cap_style(current(),Style).
set_line_cap_style(Gc,Style) -> set_item(Gc, line_cap_style, Style).

set_line_texture(Texture) ->  set_line_texture(current(),Texture).
set_line_texture(Gc,Texture) -> set_item(Gc, line_texture, Texture).

set_font(Font) -> set_font(current(), Font).
set_font(Gc,Font) -> set_item(Gc,font,Font).

%% GETTERS
%%
%% General colors
%%
get_background_color() -> get_background_color(current()).
get_background_color(Gc) -> get_item(Gc,background_color).

get_foreground_color() -> get_foreground_color(current()).
get_foreground_color(Gc) -> get_item(Gc,foreground_color).

%%
%% FILL style/color/texture
%%
get_fill_style() ->  get_fill_style(current()).
get_fill_style(Gc) -> get_item(Gc, fill_style).

get_fill_color() -> get_fill_color(current()).
get_fill_color(Gc) -> get_item(Gc, fill_color).

get_fill_texture() -> get_fill_texture(current()).
get_fill_texture(Gc) -> get_item(Gc, fill_texture).

%%
%% BORDER style/color/join/cap/texture
%%
get_border_style() ->  get_border_style(current()).
get_border_style(Gc) -> get_item(Gc, border_style).

get_border_color() -> get_border_color(current()).
get_border_color(Gc) -> get_item(Gc, border_color).

get_border_width() -> get_border_width(current()).
get_border_width(Gc) -> get_item(Gc, border_width).

get_border_join_style() -> get_border_join_style(current()).
get_border_join_style(Gc) -> get_item(Gc, border_join_style).

get_border_cap_style() -> get_border_cap_style(current()).
get_border_cap_style(Gc) -> get_item(Gc, border_cap_style).

get_border_texture() -> get_border_texture(current()).
get_border_texture(Gc) -> get_item(Gc, border_texture).

%%
%% LINE style/join/cap/texture
%%
get_line_style() ->  get_line_style(current()).
get_line_style(Gc) -> get_item(Gc, line_style).

get_line_width() -> get_line_width(current()).
get_line_width(Gc) -> get_item(Gc, line_width).

get_line_join_style() -> get_line_join_style(current()).
get_line_join_style(Gc) -> get_item(Gc, line_joine_style).

get_line_cap_style() -> get_line_cap_style(current()).
get_line_cap_style(Gc) -> get_item(Gc, line_cap_style).

get_line_texture() ->  get_line_texture(current()).
get_line_texture(Gc) -> get_item(Gc, line_texture).

get_font() -> get_font(current()).
get_font(Gc) -> get_item(Gc,font).

%%
%% General item functions
%%
set_item(Item,Value) when is_atom(Item) ->
    set_item(current(), Item, Value).

set_item(Gc, Item, Value) when is_atom(Item) ->
    epx:gc_set(Gc,Item,Value).

set(Items) ->
    set(current(), Items).

set(Gc, [{Item,Value}|Items]) ->
    ok = set_item(Gc, Item, Value),
    set(Gc, Items);
set(_Gc, []) ->
    ok.

get_item(Item) when is_atom(Item) ->
    get_item(current(), Item).

get_item(Gc, Item) when is_atom(Item) ->
    epx:gc_get(Gc, Item).

get(Items) when is_list(Items) ->
    get_item_list(current(), Items, []).
    
get(Gc, Items) when is_list(Items) ->
    get_item_list(Gc, Items, []).

get_item_list(Gc, [Item|Items], Acc) ->
    Value = get_item(Gc, Item),
    get_item_list(Gc, Items, [{Item,Value}|Acc]);
get_item_list(_Gc, [], Acc) ->
    lists:reverse(Acc).
